qiskit_ibm_runtime/runtime_session.py

```diff
@@ -0,0 +1,120 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2022.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""Qiskit runtime session."""
+
+from typing import Dict, Any, Optional, Type, Union
+from types import TracebackType
+import copy
+from functools import wraps
+
+from .ibm_runtime_service import IBMRuntimeService
+from .runtime_job import RuntimeJob
+from .runtime_program import ParameterNamespace
+from .runtime_options import RuntimeOptions
+from .program.result_decoder import ResultDecoder
+
+
+def _active_session(func):  # type: ignore
+    """Decorator used to ensure the session is active."""
+
+    @wraps(func)
+    def _wrapper(self, *args, **kwargs):  # type: ignore
+        if not self._active:
+            raise RuntimeError("The session is closed.")
+        return func(self, *args, **kwargs)
+
+    return _wrapper
+
+
+class RuntimeSession:
+    """Runtime Session"""
+
+    def __init__(
+        self,
+        runtime: IBMRuntimeService,
+        program_id: str,
+        inputs: Union[Dict, ParameterNamespace],
+        options: Optional[Union[RuntimeOptions, Dict]] = None,
+    ):
+        """RuntimeSession constructor.
+        Args:
+            runtime: Runtime service.
+            program_id: Program ID.
+            options: Runtime options.
+            inputs: Initial program inputs.
+            image: The runtime image to use, specified in the form of image_name:tag.
+        """
+        self._service = runtime
+        self._program_id = program_id
+        self._options: Optional[Union[RuntimeOptions, Dict]] = options
+        self._initial_inputs = inputs
+        self._job: Optional[RuntimeJob] = None
+        self._session_id: Optional[str] = None
+        self._active = True
+
+    @_active_session
+    def write(self, **kwargs: Dict) -> None:
+        """Write to the session."""
+        inputs = copy.copy(self._initial_inputs)
```

Comment:
The client shouldn't have to pass the initial data again. The server should be able to find the cached data.

Reply:
Fixed, now copying only for first write and not everytime.